1650 docker support

veuncent opened the pull request:
<!--- Provide a general summary of the Pull Request in the Title above -->
<!--- You can erase any part of this template that is not applicable to your Issue. -->
Initial files needed to run Arches in Docker

### Description of Change
<!--- Include a brief description of this Pull Request and reasoning behind it. -->
Docker is the world’s leading software containerization platform: https://www.docker.com/  

Some advantages of Docker:  

    - Eliminates the installation process from the end-user side: one can simply download the Arches Docker image from hub.docker.com and run the container.  
    - Keeps your system clean and provides a stable environment that is the same wherever you run it.  
    - Enhances security: the container your app is running in is isolated from the host machine.  



### Types of changes
<!--- Put an `x` in the boxes that apply  -->
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist
- [x] Unit tests pass locally with my changes 
- [x] I have added necessary documentation (if appropriate)
--> See [docker/Readme.md](docker/Readme.md)

### To-do:
- [x] Fix unit tests: Tests are not detected by `python manage.py test tests --pattern="*.py" --settings="tests.test_settings"`. Needs fixing.
- [x] Check missing time_wheel_config: Getting a 404 on: `/resource/search/time_wheel_config HTTP/1.1" 404`
- [x] Develop Readme and workflow that distinguishes between Docker for Arches core and Docker for Arches end users (custom projects).
- [x] Add remote debugging to the Image. *Note: added all that is necessary to enable remote debugging for Visual Studio Code. Added directions for this in docker/Readme.md, as well as for debugging with Pycharm.

apeters commented:
I notice you're using postgres 9.5 and mapnik 3.0 and I we're using 9.6 and 2.0 respectively

jmunowitch commented:
It looks like some versions are a bit off. This is totally our fault because we haven't updated the appropriate install scripts/documentation for this. 

Here are some updated library versions:
- Virtualenv should now be 15.1.0 (See: https://github.com/archesproject/arches/issues/1696 and https://github.com/archesproject/arches/issues/1702)
- Elasticsearch is now pinned to 5.2, it should be set to the appropriate version in the [pip install](https://github.com/archesproject/arches/pull/1748/files#diff-3254677a7917c6c01f55212f86c57fbfR96) (See: https://github.com/archesproject/arches/commit/8081c17bf401a1700efcc831a252ddef6cd21ad4)

apeters commented:
Additionally you have ip addresses hard-coded in settings_local.py.  Those should probably be removed.

veuncent commented:
Hard-coded ip addresses: these are actually services provided by AWS to check an EC2 instance's private ip (here used to check if server is running in AWS at all).

After discussion Mapnik binaries will be removed altogether. 

Other versions will be upgraded in following commit.

veuncent commented:
I pushed a number of new commits to my docker fork. Most notable changes are 
- Changing versions according to discussion
- Remove Mapnik binaries
- Removed unnecessary `pip install` commands
- Added convenient cleanup script
- Moved `setup.py develop` to entrypoint.sh, so it is only run when container starts, not during image build
- Moved init script into entrypoint.sh --> `manage.py packages -o setup` is run if psql cannot find existing db (can be forced with a flag). Please let me know what you think of this. It looks a bit scary, but it makes the whole ordeal a lot simpler. The psql command looks safe enough to me.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/archesproject/arches?pullRequest=1748) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/archesproject/arches?pullRequest=1748) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/archesproject/arches?pullRequest=1748) it.</sub>

veuncent commented:
- Updated Readme to reflect new Organization name `archesproject`
- Updated docker-compose.yml to work as production environment.
- Updated Readme to reflect this and how to use it for dev

veuncent commented:
Sure thing

apeters commented on arches/app/utils/data_management/sparql_providers/aat_provider.py:
```diff
@@ -26,7 +26,7 @@
 class AAT_Provider(Abstract_Provider):
 
     def __init__(self, **kwargs):
-        super(AAT_Provider, self).__init__('http://vocab.getty.edu/sparql.json', **kwargs)
```
This needs to be changed back to use the whole fully qualified url (eg: http://vocab.getty.edu/sparql.json)
This will fail without the protocol specified.
